ykcompile/src/lib.rs

```diff
@@ -101,15 +101,28 @@ impl<TT> CompiledTrace<TT> {
     }
 }
 
+/// Represents a memory location using a register and an offset.
+#[derive(Debug, Clone, PartialEq)]
+struct RegAndOffset {
+    reg: u8,
+    offs: i32,
+}
+
 #[derive(Debug, Clone, PartialEq)]
 enum Location {
     Register(u8),
-    Stack(i32),
-    Arg(i32),
+    Mem(RegAndOffset),
```

Comment:
Can we not merge the `struct` into the `enum` here:
```
enum Location {
    Register(u8),
    Mem { reg: u8, offs: i32 },
```

This also means we can get rid of the `new_mem` function below.

Reply:
Can do. I actually started out with something like this, but went for (too much?) abstraction.

Reply:
Less code is rarely a bad idea. :-)

Reply:
Actually, having started this change, I think I prefer the additional struct, as it makes constructing and de-structuring less pleasant. Can we keep it as-is?

Reply:
Not sure what you mean. It should be just replacing `Location::new_mem(` with `Location::Mem {`. And in the matches `Location::Mem(ro1)` with `Location::Mem(reg, off)`. To me it feels like the code would becleaner without the struct.

Reply:
I think it feels a little awkward as you have to keep trace of the two fields inside.

e.g.:
```diff
-            (Location::Register(reg), Location::Mem(ro)) => {
+            (Location::Register(reg), Location::Mem{reg: reg2, offs}) => {
                 dynasm!(self.asm
-                    ; mov Rq(reg), [Rq(ro.reg) + ro.offs]
+                    ; mov Rq(reg), [Rq(reg2) + offs]
                 );
```

I much prefer having the two related items encapsulated in `ro`.

Reply:
I think I prefer this version. You could simplify it a bit more:
```
```diff
-            (Location::Register(regl), Location::Mem(ro)) => {
+            (Location::Register(regl), Location::Mem{reg, offs}) => {
                 dynasm!(self.asm
-                    ; mov Rq(regl), [Rq(ro.reg) + ro.offs]
+                    ; mov Rq(regl), [Rq(reg) + offs]
                 );
```
```

But if it really pains you that much then we can go with yours.

Reply:
It doesn't pain me, per se. I just prefer it :P